File: token-client-core/src/main/java/no/nav/security/token/support/client/core/OAuth2CacheFactory.java
package no.nav.security.token.support.client.core;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.github.benmanes.caffeine.cache.Expiry;
import no.nav.security.token.support.client.core.oauth2.OAuth2AccessTokenResponse;
import org.checkerframework.checker.nullness.qual.NonNull;

import java.util.concurrent.TimeUnit;

public class OAuth2CacheFactory {

    public static <T> Cache<T, OAuth2AccessTokenResponse> accessTokenResponseCache(long maximumSize, long skewInSeconds) {
        // Evict based on a varying expiration policy
        return Caffeine.newBuilder()
            .maximumSize(maximumSize)
            .expireAfter(evictOnResponseExpiresIn(skewInSeconds))
            .build();
    }

    private static  <T> Expiry<T, OAuth2AccessTokenResponse> evictOnResponseExpiresIn(long skewInSeconds) {
        return new Expiry<>() {
            @Override
            public long expireAfterCreate(@NonNull T key, @NonNull OAuth2AccessTokenResponse response,
                                          long currentTime) {
                long seconds = response.getExpiresIn() > skewInSeconds ?
                    response.getExpiresIn() - skewInSeconds : response.getExpiresIn();
                return TimeUnit.SECONDS.toNanos(seconds);
            }

            @Override
            public long expireAfterUpdate(@NonNull T key, @NonNull OAuth2AccessTokenResponse response,
                                          long currentTime, long currentDuration) {
                return currentDuration;
            }

            @Override
            public long expireAfterRead(@NonNull T key, @NonNull OAuth2AccessTokenResponse response, long currentTime
                , long currentDuration) {
                return currentDuration;
            }
        };
    }
}

File: token-client-core/src/main/java/no/nav/security/token/support/client/core/OAuth2CacheFactory.kt
package no.nav.security.token.support.client.core

import com.github.benmanes.caffeine.cache.Caffeine
import com.github.benmanes.caffeine.cache.Expiry
import java.util.concurrent.TimeUnit.SECONDS
import no.nav.security.token.support.client.core.oauth2.OAuth2AccessTokenResponse

object OAuth2CacheFactory {

    @JvmStatic
    fun <T> accessTokenResponseCache(maximumSize : Long, skewInSeconds : Long) =
        // Evict based on a varying expiration policy
        Caffeine.newBuilder()
            .maximumSize(maximumSize)
            .expireAfter(evictOnResponseExpiresIn<Any>(skewInSeconds))
            .build<T, OAuth2AccessTokenResponse>()

    private fun <T> evictOnResponseExpiresIn(skewInSeconds : Long) : Expiry<T, OAuth2AccessTokenResponse> {
        return object : Expiry<T, OAuth2AccessTokenResponse> {
            override fun expireAfterCreate(key : T, response : OAuth2AccessTokenResponse, currentTime : Long) =
                SECONDS.toNanos(if (response.expiresIn > skewInSeconds) response.expiresIn - skewInSeconds else response.expiresIn.toLong())

            override fun expireAfterUpdate(key : T, response : OAuth2AccessTokenResponse, currentTime : Long, currentDuration : Long) = currentDuration
            override fun expireAfterRead(key : T, response : OAuth2AccessTokenResponse, currentTime : Long, currentDuration : Long) = currentDuration
        }
    }
}